lib/core/src/Cardano/Wallet.hs

```diff
@@ -917,23 +919,32 @@ restoreWallet ctx wid = db & \DBLayer{..} -> do
 -- | Rewind the UTxO snapshots, transaction history and other information to a
 -- the earliest point in the past that is before or is the point of rollback.
 rollbackBlocks
-    :: forall ctx s k. (HasDBLayer IO s k ctx)
+    :: forall ctx s k.
+        ( HasDBLayer IO s k ctx
+        , HasGenesisData ctx
+        )
     => ctx
     -> WalletId
     -> ChainPoint
     -> ExceptT ErrNoSuchWallet IO ChainPoint
 rollbackBlocks ctx wid point = db & \DBLayer{..} -> do
-    mapExceptT atomically $ toChainPoint <$> rollbackTo wid (pseudoPointSlot point)
+    mapExceptT atomically $ (toChainPoint block0)
+        <$> rollbackTo wid (pseudoPointSlot point)
   where
     db = ctx ^. dbLayer @IO @s @k
+    (block0, _, _) = ctx ^. genesisData
 
--- See NOTE [PointSlotNo]
-pseudoPointSlot :: ChainPoint -> SlotNo
-pseudoPointSlot ChainPointAtGenesis = W.SlotNo 0
-pseudoPointSlot (ChainPoint slot _) = slot
+    -- See NOTE [PointSlotNo]
+    pseudoPointSlot :: ChainPoint -> SlotNo
+    pseudoPointSlot ChainPointAtGenesis = W.SlotNo 0
+    pseudoPointSlot (ChainPoint slot _) = slot
 
-toChainPoint :: W.BlockHeader -> ChainPoint
-toChainPoint (BlockHeader slot _ h _) = ChainPoint slot h
+toChainPoint :: W.Block -> W.BlockHeader -> ChainPoint
```

Comment:
An alternative to `W.Block` would be take a `W.Hash "Genesis"` directly, which is a type we use in a few places:

```
lib/core/src/Cardano/Wallet.hs:    | ErrCheckIntegrityDifferentGenesis (Hash "Genesis") (Hash "Genesis")
lib/core/src/Cardano/Wallet/Api/Types.hs:    { genesisBlockHash :: !(ApiT (Hash "Genesis"))
lib/core/src/Cardano/Wallet/Api/Types.hs:instance FromJSON (ApiT (Hash "Genesis")) where
lib/core/src/Cardano/Wallet/Api/Types.hs:instance ToJSON (ApiT (Hash "Genesis")) where
lib/core/src/Cardano/Wallet/Primitive/Types.hs:    { getGenesisBlockHash :: Hash "Genesis"
lib/core/src/Cardano/Wallet/Primitive/Types/Hash.hs:instance FromText (Hash "Genesis")         where fromText = hashFromText 32
lib/core/test-common/Cardano/Wallet/DummyTarget/Primitive/Types.hs:genesisHash :: Hash "Genesis"
lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:            jsonRoundtripAndGolden $ Proxy @(ApiT (Hash "Genesis"))
lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:            |] `shouldBe` (Left @String @(ApiT (Hash "Genesis")) msg)
lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:instance Arbitrary (Hash "Genesis") where
lib/core/test/unit/Cardano/Wallet/Primitive/SlottingSpec.hs:instance Arbitrary (Hash "Genesis") where
lib/core/test/unit/Cardano/Wallet/Primitive/SlottingSpec.hs:    arbitrary = return $ Hash "Genesis Hash"
lib/core/test/unit/Cardano/Wallet/Primitive/Types/HashSpec.hs:        textRoundtrip $ Proxy @(Hash "Genesis")
lib/core/test/unit/Cardano/Wallet/Primitive/Types/HashSpec.hs:            fromText @(Hash "Genesis") "----" ===
lib/core/test/unit/Cardano/Wallet/Primitive/Types/HashSpec.hs:instance Arbitrary (Hash "Genesis") where
lib/shelley/src/Cardano/Wallet/Byron/Compatibility.hs:fromChainHash :: W.Hash "Genesis" -> ChainHash ByronBlock -> W.Hash "BlockHeader"
lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:    => W.Hash "Genesis"
lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:    :: W.Hash "Genesis"
lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:    :: W.Hash "Genesis"
lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:toTip :: W.Hash "Genesis" -> W.BlockHeader -> Tip (CardanoBlock sc)
lib/shelley/test/unit/Cardano/Wallet/Shelley/CompatibilitySpec.hs:instance Arbitrary (Hash "Genesis") where
```
and should prevent callers passing in just any block.

Not that important though, so feel free to keep it as is.

Reply:
I would indeed prefer to keep it as it is — my main motivation for using the block as opposed to the hash was to not duplicate the line

```
genesisHash = genesisBlock ^. (#header . #headerHash)
```

More importantly, I expect the `toChainPoint` function to be thrown out of `Cardano.Wallet` module again, and incorporated into the DB layer.